library/Sexp/src/Juvix/Sexp/Structure/Frontend.hs

```diff
@@ -1158,127 +1174,13 @@ toVia :: Sexp.T -> Maybe Via
 toVia form
   | isVia form =
     case form of
-      _ Sexp.:> hand Sexp.:> prog Sexp.:> Sexp.Nil ->
-        Via hand prog |> Just
-      _ -> Nothing
-  | otherwise = Nothing
+      _nameVia Sexp.:> sexp1 Sexp.:> sexp2 Sexp.:> Sexp.Nil ->
+        Via sexp1 sexp2 |> Just
+      _ ->
+        Nothing
+  | otherwise =
+    Nothing
 
 fromVia :: Via -> Sexp.T
-fromVia (Via handler prog) =
-  Sexp.list [Sexp.atom nameVia, handler, prog]
-
-----------------------------------------
--- Typeclass Instances
-----------------------------------------
-
-instance Structure Defun where
-  to = toDefun
-  from = fromDefun
-
-instance Structure Type where
-  to = toType
-  from = fromType
-
-instance Structure PredAns where
-  to = toPredAns
-  from = fromPredAns
-
-instance Structure Cond where
-  to = toCond
-  from = fromCond
-
-instance Structure Signature where
-  to = toSignature
-  from = fromSignature
-
-instance Structure LetSignature where
-  to = toLetSignature
-  from = fromLetSignature
-
-instance Structure LetType where
-  to = toLetType
-  from = fromLetType
-
-instance Structure Let where
-  to = toLet
-  from = fromLet
-
-instance Structure Case where
-  to = toCase
-  from = fromCase
-
-instance Structure Arrow where
-  to = toArrow
-  from = fromArrow
-
-instance Structure Lambda where
-  to = toLambda
-  from = fromLambda
-
-instance Structure NameBind where
-  to = toNameBind
-  from = fromNameBind
-
-instance Structure NotPunned where
-  to = toNotPunned
-  from = fromNotPunned
-
-instance Structure Punned where
-  to = toPunned
-  from = fromPunned
-
-instance Structure Record where
-  to = toRecord
-  from = fromRecord
-
-instance Structure Infix where
-  to = toInfix
-  from = fromInfix
-
-instance Structure Open where
-  to = toOpen
-  from = fromOpen
-
-instance Structure OpenIn where
-  to = toOpenIn
-  from = fromOpenIn
-
-instance Structure Declare where
-  to = toDeclare
-  from = fromDeclare
-
-instance Structure Declaim where
-  to = toDeclaim
-  from = fromDeclaim
-
-instance Structure DefModule where
-  to = toDefModule
-  from = fromDefModule
-
-instance Structure LetModule where
-  to = toLetModule
-  from = fromLetModule
-
-instance Structure Effect where
-  to = toEffect
-  from = fromEffect
-
-instance Structure DefHandler where
-  to = toDefHandler
-  from = fromDefHandler
-
-instance Structure LetOp where
-  to = toLetOp
-  from = fromLetOp
-
-instance Structure LetRet where
-  to = toLetRet
-  from = fromLetRet
-
-instance Structure Do where
-  to = toDo
-  from = fromDo
-
-instance Structure Via where
-  to = toVia
-  from = fromVia
+fromVia (Via sexp1 sexp2) =
+  Sexp.list [Sexp.atom nameVia, sexp1, sexp2]
```

Comment:
It would be better to just generate code that instantiate each structure than doing this.